### PR TITLE
!!! [v6r15] Allow extensions to extend DErrno

### PR DESCRIPTION
**Pull request by chaen**
This allows extensions to define their specific error codes
It also groups the error number by system for better tracking of them

### CODE REVIEW

**Comment by atsareg** [comment]
Please, do not put extra stuff into the global __init__(). This will be executed all the time in any context if even it is not expected. We should rather think of not importing anything from DIRAC but from where the corresponding components are actually defined

**Comment by chaen** [comment]
Originally I tried to do it in DErrno directly, but I could not import the ObjectLoader there. I can try again by modifying the import order though...

**Comment by chaen** [comment]
In fact, what I have now does not even work in the production setup (it worked in my tests because the ExtraModules is in the local file). I am getting lost... I could wrap up all this in a DErrno method, but it needs to be called once we have access to the CS. And after one and half year, I still don't know how/when/when it is all initialized. Can someone educate me ?

**Comment by rajanandakumar** [comment]
Are you by chance missing the parceCommandLine bit, like for example in

https://twiki.cern.ch/twiki/bin/view/Main/DirtyDiracTricks#Quick_and_dirty_testing

**Comment by chaen** [comment]
This parseCommandLine (at least from what I understood) is only for interactive use like scripts. Not for the code itself, or the Agent and so on. But I am sure there is "something" doing pretty much the same, but it is "somewhere" :-)

**Comment by rajanandakumar** [comment]
I am not sure about what exactly you are doing, but maybe Ricardo's old comment here may help

https://github.com/DIRACGrid/DIRAC/issues/1338#issuecomment-14769618

But then this is the limit of my knowledge on this, except for of course

http://diracgrid.org/files/docs/DeveloperGuide/AddingNewComponents/DevelopingCommands/index.html

**Comment by chaen** [comment]
Thanks Raja, I did not see this before :-) . It is indeed interesting, but still targets only scripts :-( And I assume he never wrote the explanation Andrei asked for. 
FYI, what I am trying to do is to get the list of extension as soon as possible to update the internals of the DError module before it actually get used by extensions.

**Comment by rajanandakumar** [comment]
From my point of view - whatever works. But then that is why I suppose I am not really a proper developer => sorry I cannot help you here ... :)

**Comment by atsareg** [comment]
parseCommandLine() in the scripts activates the DIRAc configuration subsystem. No surprise that DIRAC agents, services, executors are also started by scripts. Therefore, in the scripts you have an activated CS after this call and you can import whatever other modules you need. In the other components in the code written by the developers ( not in the framework ) the CS is already activated. Often people forget that for their quick tests, e.g. in the python prompt, and try to run without the CS activation.

**Comment by chaen** [comment]
If I try to call CSGlobals.getCSExtensions() at the end of the __init__.py, and I have script importing a module, be it with or without the parseCommandLine first, the extension list is empty. So my question would be "where in the code can I call CSGlobals.getCSExtensions(), make sure it is systematically called (at best once only), and make sure to get an answer"

**Comment by atsareg** [comment]
I guess you mean the DIRAC.__init__.py . The code there is executed BEFORE anything else is imported, including parseCommandLine. So, no surprise you get there an empty list. That's why I asked you not to put any non-trivial initialization code into DIRAC.__init__.py

File: DataManagementSystem/Client/DataManager.py [hunk 1]
```diff
@@ -16,7 +16,8 @@
 import fnmatch, os, time
 # # from DIRAC
 import DIRAC
-from DIRAC import S_OK, S_ERROR, DError, DErrno, gLogger, gConfig
+from DIRAC import S_OK, S_ERROR, gLogger, gConfig
+from DIRAC.Core.Utilities import DErrno, DError
```
Comment (fstagni) [inline DataManagementSystem/Client/DataManager.py]: Isn't it 

from DIRAC.Core.Utilities.DErrno import DError?

**Comment by fstagni** [comment]
@rajanandakumar : regarding Ricardo's comment that you mentioned: what he was saying about the DIRAC APIs is NOT anymore true since long time.

File: DataManagementSystem/Client/DataManager.py (hunk 1 above)
Comment (chaen) [inline DataManagementSystem/Client/DataManager.py]: No, I made the DError object importable directly from the Utilities module to avoid 2 lines everywhere

**Comment by fstagni** [comment]
The guide for using DError should be updated http://diracgrid.org/files/docs/DeveloperGuide/AddingNewComponents/Utilities/DErrno/index.html

**Comment by chaen** [comment]
You mean with respect to the import ? 
https://github.com/DIRACGrid/DIRACDocs/pull/62

File: DataManagementSystem/Client/DataManager.py (hunk 1 above)
Comment (fstagni) [inline DataManagementSystem/Client/DataManager.py]: Ah alright I see

**Comment by atsareg** [comment]
Conflicts unfortunately, please fix

**Comment by atsareg** [comment]
Conflicts, please resolve

**Comment by chaen** [comment]
Anyway it still does not work because of that:
File "/tmp/loader.py", line 1, in <module>
    from DIRAC.Core.Base.Script import parseCommandLine
File "/home/chaen/dirac/DIRAC/__init__.py", line 118, in <module>
    from DIRAC.Core.Utilities.Network import getFQDN
File "/home/chaen/dirac/DIRAC/Core/Utilities/__init__.py", line 9, in <module>
    from DIRAC.Core.Utilities.Os                 import uniquePath, getDiskSpace, getDirectorySize, sourceEnv, unifyLdLibraryPath, DEBUG
File "/home/chaen/dirac/DIRAC/Core/Utilities/Os.py", line 13, in <module>
    from DIRAC.Core.Utilities.Subprocess import shellCall
File "/home/chaen/dirac/DIRAC/Core/Utilities/Subprocess.py", line 48, in <module>
    from DIRAC.Core.Utilities import DEncode
File "/home/chaen/dirac/DIRAC/Core/Utilities/DEncode.py", line 20, in <module>
    from DIRAC.Core.Utilities.DErrno import DError


I am slowly starting to loose my nerves on this. There must be a place where we are *sure* that the CS is loaded and we can just say "loadDErrnoExtensions" no ??

**Comment by chaen** [comment]
I keep moving the faulty imports inside the methods/functions, but it won't help, simply because BaseTransport imports DEncode which imports DErrno. And since we need BaseTransport to contact the CS first, we are doomed... so the very only option is to encapsulate this error merging into a function of DErrno, which will be called once we can have the extension list. But where is this bloody place ? :-)

**Pull request closed without merging**